relentless/optimize/objective.py

```diff
@@ -165,3 +181,150 @@ def design_variables(self):
+        self.communicator = communicator
+
+    def compute(self, directory=None):
+        """Evaluate the value and gradient of the relative entropy function.
+
+
+        Parameters
+        ----------
+        directory : :class:`~relentless.data.Directory`
+            The directory to which the values of the pair potential design
+            variables at the time of computation are saved (defaults to ``None``).
+
+        Returns
+        -------
+        :class:`ObjectiveFunctionResult`
+            The result of :meth:`ObjectiveFunction.make_result()`
+
+        """
+        sim = self.simulation.run(self.target, self.potentials, directory, self.communicator)
+        sim_ens = self.thermo.extract_ensemble(sim)
+
+        # compute the relative entropy gradient by integration
+        g_tgt = self.target.rdf
+        g_sim = sim_ens.rdf
+        dvars = self.design_variables()
+        gradient = _collections.KeyedArray(keys=dvars)
+
+        for var in dvars:
+            update = 0
+            for i,j in self.target.pairs:
+                rs = self.potentials.pair.r
+                us = self.potentials.pair.energy((i,j))
+                dus = self.potentials.pair.derivative((i,j),var)
+
+                #only count (continuous range of) finite values
+                flags = np.isinf(us)
+                first_finite = 0
+                while not flags[first_finite] and first_finite < len(rs):
+                    first_finite += 1
+                rs = rs[first_finite+1:]
+                dus = dus[first_finite+1:]
+
+                #interpolate derivative wrt design variable with r
+                dudvar = _math.Interpolator(rs,dus)
+
+                # find common domain to compare rdfs
+                r0 = max(g_sim[i,j].table[0][0],g_tgt[i,j].table[0][0])
+                r1 = min(g_sim[i,j].table[-1][0],g_tgt[i,j].table[-1][0],self.potentials.pair.rmax)
+                r = np.arange(r0,r1+0.5*self.dr,self.dr)
+
+                # take integral by trapezoidal rule
+                sim_factor = sim_ens.N[i]*sim_ens.N[j]*sim_ens.beta/(sim_ens.V.volume*self.target.V.volume)
+                tgt_factor = self.target.N[i]*self.target.N[j]*self.target.beta/(self.target.V.volume**2)
+                mult = 1 if i == j else 2 # 1 if same, otherwise need i,j and j,i contributions
+                y = 2*mult*np.pi*r**2*(sim_factor*g_sim[i,j](r)-tgt_factor*g_tgt[i,j](r))*dudvar(r)
+                update += scipy.integrate.trapz(x=r, y=y)
+
+            gradient[var] = update
+
+        # optionally write output to directory
+        if directory is not None:
+            if self.communicator is None or self.communicator.rank == self.communicator.root:
+                n = 0
+                for p in self.potentials.pair.potentials:
+                    fname = 'potential_{}.log'.format(n)
+                    with open(directory.file(fname),'w') as f:
+                        p.save(f.name)
+                    n += 1
```

Comment:
This is mostly correct but can be simplified (see below). Our only big issue is that we don't have an obvious mapping between potentials and parameters (like, how do you know what object is 0 or 1), but this is not easy to solve here. Let's just stick with this for now:
```suggestion
        if directory is not None and (self.communicator is None or self.communicator.rank == self.communicator.root):
            for n,p in enumerate(self.potentials.pair.potentials):
                p.save(directory.file('potential.{}.json'.format(n)))
```